File: 1383-maximum-performance-of-a-team/1383-maximum-performance-of-a-team.java
class Solution {
    
    public static final int MOD = 1000000007;
    
    public int maxPerformance(int n, int[] speed, int[] efficiency, int k) {    
        
        int[][] arr = new int[n][2];
        for(int i=0;i<n;i++) {
            arr[i][0]=efficiency[i];
            arr[i][1]=speed[i];
        }
        Arrays.sort(arr, (o1, o2) -> (o2[0] - o1[0]));
        
        PriorityQueue<Integer> q = new PriorityQueue<>();
        long sum = 0, ans = 0;
        for(int i = 0; i < n; i++) {
            sum += arr[i][1];
            q.add(arr[i][1]);
            if (q.size() > k) sum -= q.poll();
            ans = Math.max(ans,sum*arr[i][0]);
        }
        return (int) (ans % MOD);
    }
}

File: 1383-maximum-performance-of-a-team/1383-maximum-performance-of-a-team.py
class Solution:
    def maxPerformance(self, n: int, s: List[int], e: List[int], k: int) -> int:
        se = [(-e[i], s[i]) for i in range(n)]
        h = []
        heapify(se)
        MOD = 10**9+7
        curr, ans = 0, 0
        for i in range(n):
            eff, spd = heappop(se)
            if len(h) < k:
                curr += spd
                heappush(h, spd)
                ans = max(ans, -eff*curr)
            else:
                popped = heappop(h)
                curr -= popped
                curr += spd
                ans = max(ans, -eff*curr)
                curr -= spd
                curr += max(spd, popped)
                heappush(h, max(spd, popped))
            
        return ans % MOD